File: src/layerBias.zig
const std = @import("std");

pub fn Layer(
    comptime inputSize: usize,
    comptime outputSize: usize,
) type {
    std.debug.assert(inputSize != 0);
    std.debug.assert(outputSize != 0);
    //std.debug.assert(batchSize != 0);

    return struct {
        weights: []f64,
        biases: []f64,
        last_inputs: []const f64,
        outputs: []f64,
        weight_grads: []f64,
        bias_grads: []f64,
        input_grads: []f64,
        batchSize: usize,

        const Self = @This();

        pub fn setWeights(self: *Self, weights: []f64) void {
            self.weights = weights;
        }

        pub fn setBiases(self: *Self, biases: []f64) void {
            self.biases = biases;
        }

        pub fn init(
            alloc: std.mem.Allocator,
            batchSize: usize,
        ) !Self {
            var weights: []f64 = try alloc.alloc(f64, inputSize * outputSize);
            var biases: []f64 = try alloc.alloc(f64, outputSize);
            var prng = std.rand.DefaultPrng.init(123);

            var w: usize = 0;
            while (w < inputSize * outputSize) : (w += 1) {
                weights[w] = prng.random().floatNorm(f64) * 0.2;
            }

            var b: usize = 0;
            while (b < outputSize) : (b += 1) {
                biases[b] = prng.random().floatNorm(f64) * 0.2;
            }

            return Self{
                .weights = weights,
                .biases = biases,
                .last_inputs = undefined,
                .outputs = try alloc.alloc(f64, outputSize * batchSize),
                .weight_grads = try alloc.alloc(f64, inputSize * outputSize),
                .bias_grads = try alloc.alloc(f64, outputSize),
                .input_grads = try alloc.alloc(f64, inputSize * batchSize),
                .batchSize = batchSize,
            };
        }

        pub fn forward(
            self: *Self,
            inputs: []const f64,
        ) void {
            std.debug.assert(inputs.len == inputSize * self.batchSize);

            var b: usize = 0;
            while (b < self.batchSize) : (b += 1) {
                var o: usize = 0;
                while (o < outputSize) : (o += 1) {
                    var sum: f64 = 0;
                    var i: usize = 0;
                    while (i < inputSize) : (i += 1) {
                        sum += inputs[b * inputSize + i] * self.weights[outputSize * i + o];
                    }
                    self.outputs[b * outputSize + o] = sum + self.biases[o];
                }
            }
            self.last_inputs = inputs;
        }

        pub fn backwards(
            self: *Self,
            grads: []f64,
        ) void {
            std.debug.assert(self.last_inputs.len == inputSize * self.batchSize);

            @memset(self.input_grads, 0);
            @memset(self.weight_grads, 0);
            @memset(self.bias_grads, 0);

            var b: usize = 0;
            while (b < self.batchSize) : (b += 1) {
                var o: usize = 0;
                while (o < outputSize) : (o += 1) {
                    self.bias_grads[o] += grads[b * outputSize + o] / @as(f64, @floatFromInt(self.batchSize));
                    var i: usize = 0;
                    while (i < inputSize) : (i += 1) {
                        self.weight_grads[i * outputSize + o] +=
                            (grads[b * outputSize + o] * self.last_inputs[b * inputSize + i]) / @as(f64, @floatFromInt(self.batchSize));
                        self.input_grads[b * inputSize + i] +=
                            grads[b * outputSize + o] * self.weights[i * outputSize + o];
                    }
                }
            }
        }

        pub fn applyGradients(self: *Self, weight_grads: []f64, bias_grads: []f64) void {
            var i: usize = 0;
            while (i < inputSize * outputSize) : (i += 1) {
                self.weights[i] -= 0.01 * weight_grads[i];
            }

            var o: usize = 0;
            while (o < outputSize) : (o += 1) {
                self.biases[o] -= 0.01 * bias_grads[o];
            }
        }
    };
}

File: src/main/java/LayerBias.java
import java.util.Random;

public class LayerBias {
    private double[] weights;
    private double[] biases;
    private double[] lastInputs;
    private double[] outputs;
    private double[] weightGrads;
    private double[] biasGrads;
    private double[] inputGrads;
    private int batchSize;
    private int inputSize;
    private int outputSize;

    public LayerBias(int batchSize, int inputSize, int outputSize) {
        this.batchSize = batchSize;
        this.inputSize = inputSize;
        this.outputSize = outputSize;
        this.weights = new double[inputSize * outputSize];
        this.biases = new double[outputSize];
        this.outputs = new double[outputSize * batchSize];
        this.weightGrads = new double[inputSize * outputSize];
        this.biasGrads = new double[outputSize];
        this.inputGrads = new double[inputSize * batchSize];
        Random prng = new Random(123);

        for (int i = 0; i < inputSize * outputSize; i++) {
            weights[i] = prng.nextGaussian() * 0.2;
        }

        for (int i = 0; i < outputSize; i++) {
            biases[i] = prng.nextGaussian() * 0.2;
        }
    }

    public void setWeights(double[] weights) {
        this.weights = weights;
    }

    public void setBiases(double[] biases) {
        this.biases = biases;
    }

    public void readParams(double[] params) {
        System.arraycopy(params, 0, this.weights, 0, this.weights.length);
        System.arraycopy(params, this.weights.length, this.biases, 0, this.biases.length);
    }

    public void writeParams(double[] params) {
        System.arraycopy(this.weights, 0, params, 0, this.weights.length);
        System.arraycopy(this.biases, 0, params, this.weights.length, this.biases.length);
    }

    public void deinitBackwards() {
        // No need to free memory in Java
    }

    public void forward(double[] inputs) {
        if (inputs.length != this.inputSize * this.batchSize) {
            System.out.printf("size mismatch %d, vs expected %d * %d = %d", inputs.length, this.inputSize, this.batchSize, this.inputSize * this.batchSize);
        }
        assert inputs.length == this.inputSize * this.batchSize;

        for (int b = 0; b < this.batchSize; b++) {
            for (int o = 0; o < this.outputSize; o++) {
                double sum = 0;
                for (int i = 0; i < this.inputSize; i++) {
                    sum += inputs[b * this.inputSize + i] * this.weights[this.outputSize * i + o];
                }
                this.outputs[b * this.outputSize + o] = sum + this.biases[o];
            }
        }
        this.lastInputs = inputs;
    }

    public void backwards(double[] grads) {
        assert this.lastInputs.length == this.inputSize * this.batchSize;

        for (int i = 0; i < this.inputGrads.length; i++) {
            this.inputGrads[i] = 0;
        }
        for (int i = 0; i < this.weightGrads.length; i++) {
            this.weightGrads[i] = 0;
        }
        for (int i = 0; i < this.biasGrads.length; i++) {
            this.biasGrads[i] = 0;
        }

        for (int b = 0; b < this.batchSize; b++) {
            for (int o = 0; o < this.outputSize; o++) {
                this.biasGrads[o] += grads[b * this.outputSize + o] / (double) this.batchSize;
                for (int i = 0; i < this.inputSize; i++) {
                    this.weightGrads[i * this.outputSize + o] += (grads[b * this.outputSize + o] * this.lastInputs[b * this.inputSize + i]) / (double) this.batchSize;
                    this.inputGrads[b * this.inputSize + i] += grads[b * this.outputSize + o] * this.weights[i * this.outputSize + o];
                }
            }
        }
    }

    public void applyGradients() {
        for (int i = 0; i < this.inputSize * this.outputSize; i++) {
            this.weights[i] -= 0.01 * this.weightGrads[i];
        }

        for (int o = 0; o < this.outputSize; o++) {
            this.biases[o] -= 0.01 * this.biasGrads[o];
        }
    }
}
